README.md
![Spacetime Logo](https://raw.githubusercontent.com/JoelBesada/spacetime/master/assets/Logo.png)
# Spacetime - Automatic Workspace Time Tracking
Spacetime is a VSCode extension that automatically tracks time spent in different workspaces (projects) and presents your daily activity in a nice, interactive chart. 

![Spacetime Screenshot](https://raw.githubusercontent.com/JoelBesada/spacetime/master/assets/Screenshot.png)

## Features

- Set it and forget it, Spacetime is a fully automatic time tracker. 
- Spacetime visualizes how much time you've put into each of your projects, on a daily, weekly, monthly and yearly basis. 
- Come back and see exactly how much time you sunk into all those projects that you never ended up finishing. 

## Commands
| Name | Description |
| ---- | ----------- |
| `Spacetime: View Stats` | Open up the stats panel for viewing time spent in each workspace. |


## Extension Settings
| Name | Description |
| ---- | ----------- |
| `spacetime.maxIdleMinutes` | Spacetime tracks work activity by looking at time elapsed between file save events within each workspace. This setting defines the maximum number of minutes that spacetime considers as active work time prior to a new save event occuring.|


## Release Notes

### 1.0.0

Initial release of Spacetime


CHANGELOG.md
# Change Log

All notable changes to the "spacetime" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/inlinedScript.ts
// eslint-disable-next-line @typescript-eslint/naming-convention
export default (document: any, window: any, Chart: any) => {
  const CHART_COLORS = [
    "rgb(250, 211, 144)",
    "rgb(248, 194, 145)",
    "rgb(106, 137, 204)",
    "rgb(130, 204, 221)",
    "rgb(184, 233, 148)",
    "rgb(246, 185, 59)",
    "rgb(229, 80, 57)",
    "rgb(74, 105, 189)",
    "rgb(96, 163, 188)",
    "rgb(120, 224, 143)",
    "rgb(250, 152, 58)",
    "rgb(235, 47, 6)",
    "rgb(30, 55, 153)",
    "rgb(60, 99, 130)",
    "rgb(56, 173, 169)",
    "rgb(229, 142, 38)",
    "rgb(183, 21, 64)",
    "rgb(12, 36, 97)",
    "rgb(10, 61, 98)",
  ];

  const formatTime = (_time: number) => {
    const time = Math.round(_time);
    const hours = Math.floor(time / 60 / 60);
    const minutes = Math.floor((time - hours * 60 * 60) / 60); 
    const seconds = time % 60;
    const minutesString = `${minutes} minute${minutes > 1 ? 's' : ''}`;
    if (time < 60) {return '< 1 minute';};
    if (hours < 1) {return minutesString;}
    return `${hours} hour${hours > 1 ? 's' : ''}${minutes ? `, ${minutesString}`: ''}`;
  };

  let currentChart: any;
  const renderChart = (startDate: string, endDate: string, group: 'daily' | 'weekly' | 'monthly' | 'yearly') => {
    const startTimestamp = +new Date(startDate);
    const endTimestamp = +new Date(endDate);
    const timestamps: number[] = [];

    let currentTimestamp = startTimestamp;
    while (currentTimestamp <= endTimestamp) {
      timestamps.push(currentTimestamp);
      currentTimestamp += 1000 * 60 * 60 * 24;
    }

    const buckets = timestamps.reduce<Record<string, boolean>>((acc, time) => {
      const date = new Date(time).toISOString().split('T')[0];
      const [year, month, day] = date.split('-');
      const groupDate = (() => {
        if (group === 'daily') {return [year, month, day].join('-');};
        if (group === 'weekly') {
          // The week starts on a Monday and I will not be convinced otherwise
          const weekday = (new Date(time).getDay() + 6) % 7;
          return new Date(time - weekday * 1000 * 60 * 60 * 24).toISOString().split('T')[0];
        };
        if (group === 'monthly') {return [year, month].join('-');};
        return year;
      })();
      acc[groupDate] = true;
      return acc;
    }, {});

    const data = {
      labels: Object.keys(buckets).map((bucket) => {
        const isCurrentYear = new Date(bucket).getFullYear() === new Date().getFullYear();
        return (group === 'weekly' ? 'Week of ' : '') + new Date(bucket).toLocaleString('en-US', {
          day: group === 'daily' || group === 'weekly' ? 'numeric' : undefined, 
          month: group === 'yearly' ? undefined : 'short', 
          year: isCurrentYear && group !== 'yearly' ? undefined : 'numeric'
        });
      }),
      datasets: Object.keys(window.workspaceTimes).map((workspace, index) => {
        return {
          label: workspace,
          data: Object.keys(buckets).map((bucket) => {
            return timestamps
              .filter(timestamp => {
                if (group === 'weekly') {
                  const weekday = (new Date(timestamp).getDay() + 6) % 7;
                  return new Date(timestamp - weekday * 1000 * 60 * 60 * 24).toISOString().split('T')[0].indexOf(bucket) === 0;
                }
                return new Date(timestamp).toISOString().split('T')[0].indexOf(bucket) === 0;
              })
              .reduce((acc, timestamp) => {
                acc += window.workspaceTimes[workspace][new Date(timestamp).toISOString().split('T')[0]] || 0;
                return acc;
              }, 0);
          }),
          backgroundColor: CHART_COLORS[(index * 2) % CHART_COLORS.length]
        };
      })
    };
  
    const config = {
      type: 'bar',
      data,
      options: {
        responsive: true,
        scales: {
          x: {
            stacked: true,
          },
          y: {
            stacked: true,
            ticks: {
              callback: (value: number) => {
                if (!value) {return null;};
                return formatTime(Math.floor(value / 60 / 60) * 60 * 60);
              }
            }
          }
        },
        plugins: {
          tooltip: {
            callbacks: {
              label: (tooltipItem: any) =>{
                if (!tooltipItem.raw) {
                  return null;
                };
                return `${tooltipItem.dataset.label}: ${formatTime(tooltipItem.raw)}`;
              }
            }
          }
        }
      },
    };

    if (currentChart) {
      currentChart.destroy();
    }
  
    currentChart = new Chart(
      document.getElementById('chart'),
      config
    );
  };

  const renderTable = (startDate: string, endDate: string) => {
    const tbody = document.getElementById('table-body');
    const startTimestamp = +new Date(startDate);
    const endTimestamp = +new Date(endDate);
    tbody.innerHTML = '';
    Object.keys(window.workspaceTimes).forEach((workspace, index) => {
      const totalTime = Object.keys(window.workspaceTimes[workspace]).reduce((acc, date) => {
        const time = window.workspaceTimes[workspace][date];
        const timestamp = +new Date(date);
        if (timestamp < startTimestamp || timestamp > endTimestamp) {
          return acc;
        }
        return acc + time;
      }, 0);

      const tr = document.createElement('tr');
      tr.innerHTML = `
        <td>
          <div class="workspace-color" style="background-color: ${CHART_COLORS[(index * 2) % CHART_COLORS.length]}"></div>
          ${workspace}
        </td>
        <td>${formatTime(totalTime)}</td>
      `;
      tbody.appendChild(tr);
    });
  };

  const render = () => {
    const startDate = document.getElementById('start').value;
    const endDate = document.getElementById('end').value;
    const group = document.getElementById('group').value;

    renderChart(startDate, endDate, group);
    renderTable(startDate, endDate);
  };
  
  document.getElementById('start').addEventListener('change', render);
  document.getElementById('end').addEventListener('change', render);
  document.getElementById('group').addEventListener('change', render);
  render();
};

src/extension.ts
import * as vscode from 'vscode';
import * as path from 'path';
import inlinedScript from './inlinedScript';

const WORK_SPACE_TIMES_STORAGE_KEY = 'workspaceTimes';
const MAX_IDLE_TIME_SECONDS = 60 * 15;

type WorkspaceEvent = {
	timestamp: number,
	type: 'opened' | 'saved' | 'closed'
};
type WorkspaceTimes = {
	[workspaceName: string]: {
		[date: string]: number // Time in seconds
	}
};

const workspaceEvents: {
	[workspaceName: string]: WorkspaceEvent[]
} = {};

export function activate(context: vscode.ExtensionContext) {
	context.globalState.setKeysForSync([WORK_SPACE_TIMES_STORAGE_KEY]);

	vscode.workspace.workspaceFolders?.forEach(({name}) => {
		workspaceEvents[name] = [
			{timestamp: Date.now(), type: 'opened'}
		];
	});

	vscode.workspace.onDidSaveTextDocument((document) => {
		const maxIdleTimeSeconds = vscode.workspace.getConfiguration('spacetime').maxIdleMinutes * 60 || MAX_IDLE_TIME_SECONDS;
		if (document.uri.scheme === "file") {
			const savedWorkspace = vscode.workspace.getWorkspaceFolder(document.uri);
			if (savedWorkspace) {
				const {name} = savedWorkspace;
				workspaceEvents[name] = workspaceEvents[name] || [];
				const prevEvent = workspaceEvents[name][workspaceEvents[name].length - 1];
				workspaceEvents[name].push({timestamp: Date.now(), type: 'saved'});
				if (prevEvent) {
					const workTimeSeconds = Math.min(maxIdleTimeSeconds, (Date.now() - prevEvent.timestamp) / 1000);
					const workspaceTimes = context.globalState.get<WorkspaceTimes>(WORK_SPACE_TIMES_STORAGE_KEY, {});
					const date = new Date().toISOString().split('T')[0];
					workspaceTimes[name] = workspaceTimes[name] || {};
					workspaceTimes[name][date] = (workspaceTimes[name][date] || 0) + workTimeSeconds;

					context.globalState.update('workspaceTimes', workspaceTimes);
				}
			}
		}	
	});

	let disposable = vscode.commands.registerCommand('spacetime.viewStats', () => {
		const panel = vscode.window.createWebviewPanel(
			'spacetime-stats',
			'Spacetime Stats',
			vscode.ViewColumn.One,
			{
				enableScripts: true
			}
		);

		const dayInMilliseconds = 1000 * 60 * 60 * 24;
		const today = new Date(Date.now()).toISOString().split('T')[0];
		const sevenDaysAgo = new Date(Date.now() - dayInMilliseconds * 7).toISOString().split('T')[0];

		const logoURI = panel.webview.asWebviewUri(vscode.Uri.file(path.join(context.extensionPath, 'assets', 'Logo.png')));

		panel.webview.html = `<!DOCTYPE html>
		<html lang="en">
		<head>
				<meta charset="UTF-8">
				<meta name="viewport" content="width=device-width, initial-scale=1.0">
				<title>Spacetime Stats</title>
				<style>
					h1 {
						font-size: 2.5em;
					}
					h2 {
						font-size: 1.75em;
					}

					input, select {
						background: var(--vscode-input-background);
						color: var(--vscode-input-foreground);
						outline: none;
						border: none;
						box-shadow: none;
						padding: 0.5em;
					}

					select {
						appearance: none;
						padding: 0.7em 0.5em;
						min-width: 80px;
					}

					.select-wrapper {
						display: inline-block;
						position: relative;
					}
					.select-wrapper:after {
						content: '';
						display: block;
						position: absolute;
						background: var(--vscode-input-foreground);
						width: 0.7em;
						height: 0.4em;
						top: 50%;
						right: 0.7em;
						margin-top: -0.2em;
						clip-path: polygon(100% 0%, 0 0%, 50% 100%);
					}

					input[type="date"]::-webkit-calendar-picker-indicator {
						filter: invert(1);
					}

					.vscode-light input, .vscode-light select {
						border: 1px solid #ccc;
					}

					.vscode-light input[type="date"]::-webkit-calendar-picker-indicator {
						filter: none;
					}

					.vscode-high-contrast input, .vscode-high-contrast select {
						border: 1px solid white;
					}

					.heading {
						display: flex;
						flex-wrap: wrap;
						justify-content: space-between;
						align-items: center;
						margin: 20px 0;
					}

					.date-inputs input {
            margin: 0 0.25em;
					}

					.input-container {
						padding: 1em 0.5em;
					}

					.heading h1 {
						margin: 0;
					}

					.header-wrapper {
						display: flex;
						align-items: center;
					}

					.header-wrapper img {
						width: 80px;
						height: 80px;
						margin-right: 1em;
					}

					.chart-section {
						max-width: 1200px;
						padding-bottom: 40px;
					}

					thead {
						background: rgba(100, 100, 100, 0.25);
						border-bottom: 1px solid rgba(100, 100, 100, 0.5);
					}
					tbody tr {
						background: rgba(100, 100, 100, 0.10);
					}
					tbody tr:nth-child(2n) {
						background: rgba(100, 100, 100, 0.25);
					}
					th, td {
							text-align: left;
							min-width: 200px;
							padding: 1em 1em;
					}
					th:not(:last-child), td:not(:last-child) {
						border-right: 1px solid rgba(100, 100, 100, 0.5);
					}

					.vscode-light thead {
						background: rgba(100, 100, 100, 0.1);
						border-bottom: 1px solid rgba(100, 100, 100, 0.25);
					}

					.vscode-light tbody tr {
						background: rgba(100, 100, 100, 0.03);
					}
					.vscode-light tbody tr:nth-child(2n) {
						background: rgba(100, 100, 100, 0.1);
					}
			
					.vscode-light th:not(:last-child), .vscode-light td:not(:last-child) {
						border-right: 1px solid rgba(100, 100, 100, 0.25);
					}

					table {
							border-collapse: collapse;
							font-size: 16px;
							width: 100%;
							max-width: 800px;
					}

					.workspace-color {
						display: inline-block;
						width: 14px;
						height: 14px;
						margin-right: 0.25em;
						vertical-align: middle;
					}

					.totals-section {
						padding-bottom: 100px;
					}
				</style>
		</head>
		<body>
		  <div class="heading">
				<div class="header-wrapper">
					<img src="${logoURI}" />
					<h1>Spacetime Stats</h1>
				</div>	
				<div class="input-container">
					<div class="date-inputs">
						From
						<input type="date" id="start" name="start" value=${sevenDaysAgo} max=${today}>
						to
						<input type="date" id="end" name="end" value=${today} max=${today}>
						<div class="select-wrapper"><select id="group">
							<option value="daily" selected>Daily</option>
							<option value="weekly">Weekly</option>
							<option value="monthly">Monthly</option>
							<option value="yearly">Yearly</option>
						</select></div>
					</div>
				</div>
		  </div>
      <section class="chart-section">
				<canvas id="chart"></canvas>
			</section>
      <section class="totals-section">
				<h2>Totals</h2>
				<table>
				  <thead>
					  <tr>
						  <th>Workspace</th>
							<th>Time Spent</th>
						</tr>
					</thead>
					<tbody id="table-body">
					</tbody>
				</table>
			</section>

			<script src="https://cdn.jsdelivr.net/npm/chart.js@3.2.0/dist/chart.min.js"></script>
			<script>
				window.workspaceTimes = ${JSON.stringify(context.globalState.get(WORK_SPACE_TIMES_STORAGE_KEY, {}))}
			</script>
			<script>
				${inlinedScript.toString().split('\n').slice(1, -1).join('\n')}
			</script>
		</body>
		</html>`;
	});

	context.subscriptions.push(disposable);
}

